### PR TITLE
CAL-126 Filter invalid characters from derived image filenames

### PR DESCRIPTION
**Pull request by lcrosenbu**
#### What does this PR do?
Currently derived images are stored with a filename that is derived from the NITF `FTITLE` field. The `FTITLE` field may contain characters that are invalid for file system storage (e.g. `*` is invalid for Windows system files). This PR strips out invalid characters for the purposes of generating the derived image filename. 

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@bdeining @dcruver 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@coyotesqrl
@jlcsmith

#### How should this be tested?
Ingest a NITF that has a `FTITLE` that contains invalid characters (e.g. `*`) for the file system (e.g. Windows). Verify ingest was successful and derived images can be retrieved. 

#### What are the relevant tickets?
https://codice.atlassian.net/browse/CAL-126

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java [hunk 1]
```diff
@@ -208,7 +208,8 @@ private ContentItem createDerivedImage(String id, String qualifier, BufferedImag
     }
 
     private String buildDerivedImageTitle(String title, String qualifier) {
-        String rootFileName = FilenameUtils.getBaseName(title);
+        // filter non-word characters from the filename for file system storage
+        String rootFileName = FilenameUtils.getBaseName(title).replaceAll("[\\W]", "");
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: If all the characters in the `ftitle` where special characters, you'd end up with a filename of `{qualifier}-.jpg`. You should probably handle the edge case where that happens.

Should also add unit tests to check for the various names generated, though as this is a private method it will be a little painful to test.

Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: Handled the case of null or empty `FTITLE` to produce a derived filename of `{qualifier}.jpg`. Added unit tests for various permutations (title is null, empty, valid, contains invalid characters, contains all invalid characters, contains only special chars that are non-alphanum/non-human readable).

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java [hunk 2]
```diff
@@ -121,6 +124,71 @@ public void testNonNitfContent() throws PluginExecutionException, IOException {
...
+        String derivedFilename = nitfPreStoragePlugin.buildDerivedImageTitle(title, qualifier);
+
+        assertThat(StringUtils.containsNone(derivedFilename, FILESYSTEM_INVALID_CHARS), is(true));
+        assertThat(derivedFilename, is("qualifier-TooLegitToQuit.jpg"));
+    }
+
+    @Test
+    public void testBuildDerivedImageTitleIsNull() {
+        String title = null;
+        String qualifier = "QualifierOnly";
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java]: `QualifierOnly` can be a constant

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java [hunk 3]
```diff
@@ -207,9 +210,14 @@ private ContentItem createDerivedImage(String id, String qualifier, BufferedImag
         return null;
     }
 
-    private String buildDerivedImageTitle(String title, String qualifier) {
+    protected String buildDerivedImageTitle(String title, String qualifier) {
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: If this is `protected` just for testing purposes, make it package-private instead.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java [hunk 4]
```diff
@@ -207,9 +210,14 @@ private ContentItem createDerivedImage(String id, String qualifier, BufferedImag
...
         String rootFileName = FilenameUtils.getBaseName(title);
-        return String.format(DERIVED_IMAGE_FILENAME_PATTERN, qualifier, rootFileName, JPG);
+
+        return (StringUtils.isNotBlank(rootFileName) &&
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: `StringUtils.isAlphanumeric(rootFileName)` is a much simpler test for both these conditions. https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#isAlphanumeric(java.lang.CharSequence)

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java [hunk 5]
```diff
@@ -207,9 +210,14 @@ private ContentItem createDerivedImage(String id, String qualifier, BufferedImag
...
+                StringUtils.isNotBlank(rootFileName.replaceAll("[^A-Za-z0-9]", ""))) // any human readable content
+                ? String.format(DERIVED_IMAGE_FILENAME_PATTERN, qualifier,
+                        rootFileName.replaceAll(INVALID_FILENAME_CHARACTER_REGEX, ""), JPG)
+                : String.format("%s.%s", qualifier, JPG);
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: Is it a concern that you might get the same file name for multiple files or do name collisions not matter?

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java [hunk 6]
```diff
@@ -48,6 +49,8 @@
 
     private static final String GEO_NITF = "/i_3001a.ntf";
 
+    private static final String FILESYSTEM_INVALID_CHARS = "\\/:*?<>|";
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java]: `&` and `;` are also bad characters. `.` as well. I'm not sure you care for a unit test, however.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java [hunk 7]
```diff
@@ -121,6 +124,71 @@ public void testNonNitfContent() throws PluginExecutionException, IOException {
...
 
+    @Test
+    public void testBuildDerivedImageTitle() {
+        String title = "Too Legit To Quit";
+        String qualifier = "qualifier";
+
+        String derivedFilename = nitfPreStoragePlugin.buildDerivedImageTitle(title, qualifier);
+
+        assertThat(StringUtils.containsNone(derivedFilename, FILESYSTEM_INVALID_CHARS), is(true));
+        assertThat(derivedFilename, is("qualifier-TooLegitToQuit.jpg"));
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java]: Do you maybe want to downcase the filename as well?

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java [hunk 8]
```diff
@@ -121,6 +124,71 @@ public void testNonNitfContent() throws PluginExecutionException, IOException {
...
+        String qualifier = "QualifierOnly";
+
+        String derivedFilename = nitfPreStoragePlugin.buildDerivedImageTitle(title, qualifier);
+
+        assert(StringUtils.containsNone(derivedFilename, FILESYSTEM_INVALID_CHARS));
+        assertThat(derivedFilename, is("QualifierOnly.jpg"));
+    }
+
+    @Test
+    public void testbuildDerivedImageTitleAllValidSpecialChars() {
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java]: The multiple input/output tests in this class would have been a good opportunity for a Spock test or even a parameterized Junit test.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java (hunk 5 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: I was not sure about this and sent this question to @dcruver yesterday. He is also a reviewer on this PR.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java (hunk 4 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: This method should not actually check that the title is _only_ alphanumeric, but rather ensure that it _contains some_ alphanumeric values. Since this method strips out all chars that match [\W], underscores will remain. This check makes sure the filename does not contain only underscore, such that the resulting filename would be `{qualifier}-_.jpg`.

Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: Ah yes. I see that. I guess my only quibble then would be with the ternary. With that long/complex a conditional, I think it's a little harder to read than with an if/else (and you could remove the else and just return at the end).

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePluginParameterized.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+                        QUALIFIER + "-abunchof_invalid_characters.jpg"}});
+    }
+
+    @Before
+    public void setUp() {
+        nitfPreStoragePlugin = new NitfPreStoragePlugin();
+    }
+
+    @Test
+    public void test() {
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePluginParameterized.java]: Method name should be more specific

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java [hunk 10]
```diff
@@ -210,14 +211,23 @@ private ContentItem createDerivedImage(String id, String qualifier, BufferedImag
...
-        return (StringUtils.isNotBlank(rootFileName) &&
-                StringUtils.isNotBlank(rootFileName.replaceAll("[^A-Za-z0-9]", ""))) // any human readable content
-                ? String.format(DERIVED_IMAGE_FILENAME_PATTERN, qualifier,
-                        rootFileName.replaceAll(INVALID_FILENAME_CHARACTER_REGEX, ""), JPG)
-                : String.format("%s.%s", qualifier, JPG);
+        // title must contain some alphanumeric, human readable characters, or use default filename
+        if (StringUtils.isNotBlank(rootFileName)
+                && StringUtils.isNotBlank(rootFileName.replaceAll("[^A-Za-z0-9]", ""))) {
+            String strippedFilename = rootFileName.replaceAll(INVALID_FILENAME_CHARACTER_REGEX, "");
+            derivedImageTitle = String
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: You could just return directly here and skip the `derivedImageTitle`.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java [hunk 11]
```diff
@@ -210,14 +211,23 @@ private ContentItem createDerivedImage(String id, String qualifier, BufferedImag
...
         String rootFileName = FilenameUtils.getBaseName(title);
 
-        return (StringUtils.isNotBlank(rootFileName) &&
-                StringUtils.isNotBlank(rootFileName.replaceAll("[^A-Za-z0-9]", ""))) // any human readable content
-                ? String.format(DERIVED_IMAGE_FILENAME_PATTERN, qualifier,
-                        rootFileName.replaceAll(INVALID_FILENAME_CHARACTER_REGEX, ""), JPG)
-                : String.format("%s.%s", qualifier, JPG);
+        // title must contain some alphanumeric, human readable characters, or use default filename
+        if (StringUtils.isNotBlank(rootFileName)
+                && StringUtils.isNotBlank(rootFileName.replaceAll("[^A-Za-z0-9]", ""))) {
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: Should have asked this before...the only difference between this regex and the other one you're using is the underscore. Is there a reason for that? Could you just use the const here?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java [hunk 12]
```diff
@@ -210,14 +211,23 @@ private ContentItem createDerivedImage(String id, String qualifier, BufferedImag
...
+            String strippedFilename = rootFileName.replaceAll(INVALID_FILENAME_CHARACTER_REGEX, "");
+            derivedImageTitle = String
+                    .format(DERIVED_IMAGE_FILENAME_PATTERN, qualifier, strippedFilename, JPG)
+                    .toLowerCase();
+
+        } else {
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: If you return in the if block, you can just fall through and return the default formatted string here without the else.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePluginParameterized.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+    private String fExpected;
+
+    public TestPreStoragePluginParameterized(String input, String expectedOutput) {
+        fInput= input;
+        fExpected= expectedOutput;
+    }
+
+    @Parameterized.Parameters(name = "Test #{index}: Create derived image filename from \"{0}\"")
+    public static Collection<String[]> data() {
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePluginParameterized.java]: Wow. This really is a lot uglier than Spock! Works, though.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java (hunk 2 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java]: Created a `QUALIFIER` to be reused for all test cases.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java (hunk 3 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: Modified method access to package-private.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java (hunk 6 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java]: Originally was focused on Windows specific invalid characters. But this should definitely be expanded for additional invalid characters. Added `&` and `;` for a more robust unit test. Assuming a period character is actually safe for file names, but expect them to be stripped out.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java (hunk 7 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java]: Added `toLowerCase()`

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java (hunk 8 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/TestPreStoragePlugin.java]: Updated to be a JUnit parameterized test. Will definitely give Spock a try next time around!

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java (hunk 11 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfPreStoragePlugin.java]: Probably making things more complicated than they need to be. :) The `\W` regex allows underscores in the resulting filename. The check for alphanumeric verifies that after dumping all the alphanumerics, you aren't just left with just an underscore. I assumed that the `\W` could/would be updated in the future to allow more characters, but we would never want to append a bunch of special characters (albeit VALID characters) to the filename; e.g. `original-_.jpg`, or if `+` and `=` were added as valid characters, `original-_=+.jpg`.

**Comment by LinkMJB** [comment]
test this please

File: catalog/imaging/imaging-transformer-nitf/pom.xml [hunk 14]
```diff
@@ -177,6 +184,38 @@
                     </execution>
                 </executions>
             </plugin>
+            <!-- Mandatory plugins for using Spock -->
+            <plugin>
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/pom.xml]: The gmavenplus plugins shouldn't be needed; they're in the root pom.

**Comment by lcrosenbu** [comment]
Spock wins!

File: catalog/imaging/imaging-transformer-nitf/src/test/groovy/NitfPreStoragePluginTest.groovy [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    def "Building derived image filename from \"#ftitle\"" (String ftitle, String expectedFname) {
+        setup:
+            NitfPreStoragePlugin plugin = new NitfPreStoragePlugin();
+            def qualifier = "original";
+
+        when: "building a derived image filename"
+            def derivedFname = plugin.buildDerivedImageTitle(ftitle, qualifier);
+
+        then: "the derived filenames should not include invalid characters"
+            derivedFname.equals(expectedFname)
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/src/test/groovy/NitfPreStoragePluginTest.groovy]: Looks so much better now!

One Groovy tip: you can use `==` for equality tests.

**Comment by coyotesqrl** [comment]
:+1:

**Comment by LinkMJB** [comment]
ok to test

File: catalog/imaging/imaging-transformer-nitf/pom.xml [hunk 16]
```diff
@@ -109,6 +109,13 @@
             <artifactId>jai-imageio-jpeg2000</artifactId>
             <version>${jpeg2000.version}</version>
         </dependency>
+
+        <dependency>
```
Comment (coyotesqrl) [inline catalog/imaging/imaging-transformer-nitf/pom.xml]: Ugh. Can't believe I missed this last night. This should be using the new shaded jar:
```xml
        <!--Spock dependencies-->
        <dependency>
            <groupId>ddf.lib</groupId>
            <artifactId>spock-shaded</artifactId>
            <version>${ddf.version}</version>
            <scope>test</scope>
        </dependency>
```

File: catalog/imaging/imaging-transformer-nitf/src/test/groovy/NitfPreStoragePluginTest.groovy [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p/>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+import spock.lang.*
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/test/groovy/NitfPreStoragePluginTest.groovy]: Shouldn't be using the wildcard hero.

**Comment by jlcsmith** [comment]
Other than the import wildcard, LGTM 👍

File: catalog/imaging/imaging-transformer-nitf/src/test/groovy/NitfPreStoragePluginTest.groovy (hunk 17 above)
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/test/groovy/NitfPreStoragePluginTest.groovy]: Added specific imports:

**Comment by bdeining** [comment]
LGTM
